lib/versioning/VersioningRequestProcessor.ts

```diff
@@ -496,7 +496,26 @@ export default class VersioningRequestProcessor {
                 const versionIdFromMaster = masterVersion.getVersionId();
                 if (versionIdFromMaster === undefined ||
                     versionIdFromMaster >= versionId) {
+                    let value = request.value;
                     logger.debug('version to put is not older than master');
+                    // Delete the deprecated, null key for backward compatibility
+                    // to avoid storing both deprecated and new null keys.
+                    // If master null version was put with an older Cloudserver (or in compat mode),
+                    // there is a possibility that it also has a null versioned key
+                    // associated, so we need to delete it as we write the null key.
+                    // Deprecated null key gets deleted when the new CloudServer:
+                    // - updates metadata of a null master (options.isNull=true)
+                    // - puts metadata on top of a master null key (options.isNull=false)
+                    if (request.options.isNull !== undefined && // new null key behavior when isNull is defined.
+                        !masterVersion.isNull2Version() && // master does not support the new null key behavior yet.
+                        masterVersion.isNullVersion()) { // master is null
+                        const masterNullVersionId = masterVersion.getNullVersionId();
+                        // The deprecated null key is referenced in the "nullVersionId" property of the master key.
+                        if (masterNullVersionId) {
+                            const oldNullKey = formatVersionKey(key, masterNullVersionId);
```

Comment:
I chose to call "null key" the key that stores the null version under the new behavior, which key ends with a null byte, hence in this context it may be better to rename the variable for example `oldNullVersionKey` to avoid this ambiguity.